core/src/main/scala/kafka/server/BrokerServer.scala

```diff
@@ -362,16 +326,13 @@ class BrokerServer(
         config.brokerSessionTimeoutMs / 2 // KAFKA-14392
       )
       lifecycleManager.start(
-        () => metadataListener.highestMetadataOffset,
+        () => sharedServer.loader.lastAppliedOffset(),
```

Comment:
Just want to check that the semantics are the same here (since the terminology and code is slightly different). Before, we were sending `highestMetadataOffset` which is updated as we replay messages:

```scala
        _highestOffset = lastCommittedOffset.getOrElse(batch.baseOffset() + index)
        try {
          delta.replay(messageAndVersion.message())
        } catch {
```

We don't actually call the publishers until after we have replayed the whole batch.

With the new metadata loader infrastructure, we update the `lastAppliedOffset` after we have applied the whole batch and called the publishers.

Reply:
This is a good question. It is indeed a slightly different semantic, and the PR description should call that out.

As you said, previously we were not waiting for the updates we loaded to actually be published. Now we are. The intention was always to wait for the updates to actually be propagated before declaring the broker "caught up" so this is really fixing a bug. I'll make a note in the description.